flake.nix

```diff
@@ -0,0 +1,105 @@
+
+    local-packages = [
+      { name = "servant-util";         subdirectory = "servant-util"; }
+      { name = "servant-util-beam-pg"; subdirectory = "servant-util-beam-pg"; }
+    ];
+
+    # names of all local packages
+    local-packages-names = map (p: p.name) local-packages;
+
+    make-hs-pkgs = extra_dep: let
+      # source with gitignored files filtered out
+      projectSrc = pkgs.haskell-nix.haskellLib.cleanGit { name = "servant-util"; src = ./.; };
+
+      # haskell.nix does not support 'include' in package.yaml, we have to generate .cabal ourselves
+      cabalFiles = pkgs.runCommand "servant-util-cabal-files" {} ''
+        mkdir -p $out
+        ${pkgs.lib.concatMapStrings ({ name, subdirectory }: ''
+          ${pkgs.haskellPackages.hpack}/bin/hpack ${projectSrc}/${subdirectory} - > $out/${name}.cabal
+        '') local-packages}
+      '';
+
+      cabalsrc = pkgs.runCommand "src-with-cabal" {} ''
+        cp -r --no-preserve=mode ${projectSrc} $out
+        ${pkgs.lib.concatMapStrings ({ name, subdirectory }: ''
+          cp ${cabalFiles}/${name}.cabal $out/${subdirectory}
+        '') local-packages}
+      '';
+
+      src = (if extra_dep == null then cabalsrc else pkgs.runCommand "extradep-src" { } ''
+        mkdir -p $out
+        cp -r ${cabalsrc}/* .
+        substituteInPlace stack.yaml --replace "extra-deps:" "extra-deps:
+        ${extra_dep}"
+        cp -r ./* $out
+      '');
+    in pkgs.haskell-nix.stackProject {
+      # project src with .cabal files added
+      inherit src;
+      ignorePackageYaml = true;
+
+      modules = [{
+        # configuration for local packages
+        packages = pkgs.lib.genAttrs local-packages-names (packageName: {
+          ghcOptions = [ "-O0" "-Werror" ];
+          # run haddock for local packages
+          doHaddock = true;
+        });
+
+        # don't haddock dependencies
+        doHaddock = false;
+      }];
+    };
+
+    make-tests = depname: extra_dep: let
+      hs-pkgs = make-hs-pkgs extra_dep;
+      get-package-components = pkg: with pkg.components; with pkgs.lib;
+        optionals (pkg ? library) [ library library.haddock ]
+        ++ attrValues exes
+        ++ attrValues tests;
+
+      all-components = pkgs.lib.concatMap (pkg: get-package-components hs-pkgs.${pkg}) local-packages-names;
+      comps = pkgs.lib.genAttrs local-packages-names (pkg: hs-pkgs.${pkg}.components);
+
+      component-tests = pkgs.lib.fold pkgs.lib.mergeAttrs {} (builtins.catAttrs "tests" (builtins.attrValues comps));
+      testruns = pkgs.lib.mapAttrs' (name: drv: {
+        name = "${name}-${depname}";
+        value = pkgs.runCommandLocal "${name}-run" {} ''
+          ${drv}/bin/${name} && touch $out
+        '';
+      }) component-tests;
+    in builtins.listToAttrs (map (value: { name = "${value.name}-${depname}"; inherit value; }) all-components) // testruns;
+  in {
+    checks = pkgs.lib.fold pkgs.lib.mergeAttrs {} (pkgs.lib.mapAttrsToList make-tests extraDeps);
```

Comment:
Previously tests were split on per-package basis, let's keep it this way, thus they'll be two jobs: to check `servant-util` and `servant-util-beam-ng`

Reply:
alright, I thought building all derivations as tests would be a nice addition. looking at the test run, they seem to all be covered by just these 2 tests :+1:

Reply:
OTOH, these tests aren't that time-consuming, so apparently, the current approach is good as well. Tests are run sequentially and there is no mess in logs

Reply:
So feel free to ignore my initial suggestion :see_no_evil:

Reply:
flake is cleaner like this :shrug: